Inheriting from a class that doesn't exist#33334

monteshot opened the pull request:
- Changed PRICE_CODE constant from non-exist class to $this class
This class already have its own constant on 28 line.
const PRICE_CODE = self::CONFIGURED_PRICE_CODE;
The related issue https://github.com/magento/magento2/issues/33334
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->
The class that responsible for getting price value uses not valid constant.
I was remove incorrect constant and add the constant from this class that can be correctly overrided. 
### Related Pull Requests
<!-- related pull request placeholder -->

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/magento2#33334

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. On any properly created bundle product will be executed \Magento\Catalog\Pricing\Render\ConfiguredPriceBox because the 
\Magento\Catalog\Pricing\Price\ConfiguredPriceInterface::CONFIGURED_PRICE_CODE is in Magento\Bundle\Pricing\Price\Pool
2. The render mechanism will found the template app/code/Magento/Catalog/view/base/templates/product/price/configured_price.phtml that will get the class \Magento\Bundle\Pricing\Price\ConfiguredRegularPrice that inherits from \Magento\Bundle\Pricing\Price\ConfiguredPrice with the problem get value method
3. On line 18 of app/code/Magento/Catalog/view/base/templates/product/price/configured_price.phtml will be invoked code from \Magento\Bundle\Pricing\Price\ConfiguredPrice::getValue
4. If item was previously set using setItem the getValue code will throw exception
5. I change to self::PRICE_CODE and code no longer crash
### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->
Comments:
The class that inherits this class already have the Unit test
\Magento\Catalog\Test\Unit\Pricing\Price\ConfiguredPriceTest but it.
Also i wrote unit for getValue method

Normally on frontend the code with error not invokes. It seems to be fallback.
This code may crash when unit tests executing

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

monteshot committed:
magento/magento2#33334:Inheriting from a class that doesn't exist

- Changed PRICE_CODE constant from non-exist class to $this class

monteshot committed:
magento/magento2#33334:Inheriting from a class that doesn't exist

- calculateDiscount method must be point as FinalPrice to its class but in ConfiguredPrice.php this method not exist. Also in FinalPrice invokes the \Magento\Bundle\Pricing\Price\DiscountCalculator::calculateDiscount
- added unit test for method getValue

monteshot commented:
@magento run all tests

bgorski commented on app/code/Magento/Bundle/Pricing/Price/ConfiguredPrice.php:
```diff
@@ -146,10 +152,9 @@ public function getValue()
-                $this->priceInfo
-                    ->getPrice(BundleDiscountPrice::PRICE_CODE)
-                    ->calculateDiscount($configuredOptionsAmount);
+            if (!empty($this->item->getProduct())) {
```
The `getProduct()` method should always return a product instance according to the interface and its implementations. Therefore this will never be empty. Did you mean to check if the product has an ID? If so, I think this can be simplified to reduce nesting, for example to 
```
if ($this->item && $this->item->getProduct()->getId()) {
    $configuredOptionsAmount = $this->getConfiguredAmount()->getBaseAmount();
    
    return parent::getValue() + $this->discountCalculator->calculateDiscount(
        $this->item->getProduct(),
        $configuredOptionsAmount
    );
}

return parent::getValue();
```

monteshot replied:
I did it in this way because $this->item can be the instance of null|ItemInterface. And this interface also implements another models. And as this branch of code is fallback. I did not reserch so deep but i think that "if" statement need to  check only $this->item. Maybe i wrong.
@bgorski

monteshot committed:
Removed unnecessary nesting in if statements according to the interface

monteshot committed:
magento/magento2#33334:Inheriting from a class that doesn't exist

- Removed unnecessary nesting in if statements according to the interface

monteshot committed:
Merge remote-tracking branch 'origin/Inheriting_from_a_class_that_doesn't_exist#33334' into Inheriting_from_a_class_that_doesn't_exist#33334

monteshot commented:
Hello. 
What should i do with this semantic error @bgorski 
![image](https://user-images.githubusercontent.com/2836146/125241182-cda4ec00-e2f3-11eb-9fe4-73426baa9ace.png)

Thanks.

monteshot commented:
Neither module.xml nor composer.json i cannot change version because its absent in both files

monteshot committed:
magento/magento2#33334:Inheriting from a class that doesn't exist

- Removed unnecessary properties from test class

monteshot commented:
@magento run all tests

monteshot committed:
magento/magento2#33334:Inheriting from a class that doesn't exist

- Update the unit test for getter in case non-existing the item

monteshot commented:
@magento run all tests

monteshot committed:
magento/magento2#33334:Inheriting from a class that doesn't exist

- Made changes backwards compatible.

monteshot commented:
@magento run all tests

andrewbess reviewed (changes requested):
Hello @monteshot 
Thank you for your contribution.
Could you please rename your new tests?
The tests are related with Magento\Bundle\Pricing\Price\ConfiguredPrice
I think it should be named Magento\Bundle\Test\Unit\Pricing\Price\ConfiguredPriceTest.
How do you think about it?

monteshot commented:
@andrewbess my mistake. i'll change the comment of php file

monteshot committed:
magento#33334:Inheriting from a class that doesn't exist

- Renamed Test class according to the recommendations.

monteshot committed:
magento/magento2#33334:Inheriting from a class that doesn't exist

- Renamed Test class according to the recommendations.

monteshot commented:
@magento run all tests

monteshot committed:
Merge remote-tracking branch 'origin/Inheriting_from_a_class_that_doesn't_exist#33334' into Inheriting_from_a_class_that_doesn't_exist#33334

monteshot commented:
@magento run all tests

monteshot commented:
@magento run Functional Tests B2B

monteshot commented:
@magento run Adobe CLA Signed?

monteshot commented:
Unit test fail due a "Flaky test"

monteshot commented:
@magento run all tests

monteshot committed:
magento/magento2#33334:Inheriting from a class that doesn't exist

- Clarified comment for main test method

monteshot commented:
@magento run all tests

eduard13 commented on app/code/Magento/Bundle/Test/Unit/Pricing/Price/ConfiguredPriceTest.php:
```diff
@@ -0,0 +1,182 @@
+     */
+    protected $calculator;
+
+    /**
+     * @var MockObject
+     */
+    protected $priceInfo;
```
```suggestion
    /**
     * @var Base|MockObject
     */
    private $priceInfoMock;
```

eduard13 commented on app/code/Magento/Bundle/Test/Unit/Pricing/Price/ConfiguredPriceTest.php:
```diff
@@ -0,0 +1,182 @@
+    protected $basePriceValue = 100.;
+
+    /**
+     * @var MockObject
+     */
+    protected $item;
```
Since these aren't the real objects, I would name the properties with a `Mock` suffix. Also. please mark each property as private.
```suggestion
     * @var ItemInterface|MockObject
     */
    private $itemMock;
```

eduard13 commented on app/code/Magento/Bundle/Test/Unit/Pricing/Price/ConfiguredPriceTest.php:
```diff
@@ -0,0 +1,182 @@
+    protected $item;
+
+    /**
+     * @var MockObject
+     */
+    protected $product;
```
```suggestion
     * @var Product|MockObject
     */
    private $productMock;
```

eduard13 commented on app/code/Magento/Bundle/Test/Unit/Pricing/Price/ConfiguredPriceTest.php:
```diff
@@ -0,0 +1,182 @@
+    /**
+     * @var PriceCurrencyInterface|MockObject
+     */
+    protected $priceCurrencyMock;
+    /**
```
Please preserve an empty line after each property.
```suggestion
    protected $priceCurrencyMock;

    /**
```

eduard13 commented on app/code/Magento/Bundle/Pricing/Price/ConfiguredPrice.php:
```diff
@@ -48,6 +48,10 @@ class ConfiguredPrice extends CatalogPrice\FinalPrice implements ConfiguredPrice
      * @var ConfiguredPriceSelection
      */
     private $configuredPriceSelection;
+    /**
+     * @var \Magento\Bundle\Pricing\Price\DiscountCalculator
```
Please preserve an empty row and do not use FQPN. Please import the class name.
```suggestion

    /**
     * @var DiscountCalculator
```

eduard13 commented on app/code/Magento/Bundle/Test/Unit/Pricing/Price/ConfiguredPriceTest.php:
```diff
@@ -0,0 +1,182 @@
+        $this->model->setItem($this->item);
+    }
+
+    private function prepareAndReturnSelectionPriceDataStub()
```
Please move the private method after the public ones. Also provide the method's docblock with a short description.
```suggestion
    private function prepareAndReturnSelectionPriceDataStub(): array
```

eduard13 commented on app/code/Magento/Bundle/Test/Unit/Pricing/Price/ConfiguredPriceTest.php:
```diff
@@ -0,0 +1,182 @@
+    /**
+     * Test of value getter when item presented
+     */
+    public function testGetValueMethod()
```
```suggestion
    public function testGetValueMethod(): void
```

eduard13 commented on app/code/Magento/Bundle/Test/Unit/Pricing/Price/ConfiguredPriceTest.php:
```diff
@@ -0,0 +1,182 @@
+    /**
+     * Test of value getter if no product item
+     */
+    public function testGetValueMethodNoItem()
```
```suggestion
    public function testGetValueMethodNoItem(): void
```

eduard13 commented on app/code/Magento/Bundle/Test/Unit/Pricing/Price/ConfiguredPriceTest.php:
```diff
@@ -0,0 +1,182 @@
+    {
+        unset($this->item);
+        $this->product = $this->getMockBuilder(Product::class)
+            //->setMethods(['getPriceInfo', 'getOptionById', 'getResource', 'getId'])
```
I believe we can remove the commented out line.

eduard13 commented on app/code/Magento/Bundle/Test/Unit/Pricing/Price/ConfiguredPriceTest.php:
```diff
@@ -0,0 +1,182 @@
+     */
+    public function testGetValueMethodNoItem()
+    {
+        unset($this->item);
```
Could you please clarify the reason of removing and re-creating the `$this->item` object? Usually for a better structure and clearance, the needed objects should be instantiated within the test case.

eduard13 commented on app/code/Magento/Bundle/Test/Unit/Pricing/Price/ConfiguredPriceTest.php:
```diff
@@ -0,0 +1,182 @@
+    protected function setUp(): void
+    {
+        $basePrice = $this->getMockForAbstractClass(PriceInterface::class);
+        $basePrice->expects($this->any())->method('getValue')->willReturn($this->basePriceValue);
```
Basically the `setUp` method, is used to mock the needed objects, that will be used later by different test cases. Please move all the assertions into a separate test case instead.

monteshot replied:
Ok. But in setUp I dont have assertions

eduard13 reviewed (changes requested):
Hi @monteshot, thank you for your contribution. Could you please review the below suggestions? Please let me know if there are any concerns or questions.

Thank you.

monteshot committed:
magento/magento2#33334:Inheriting from a class that doesn't exist

- Adjusted code style for the unit test
- Replaced FQPN using import

monteshot commented:
@magento run all tests

monteshot commented:
@magento run Functional Tests EE

monteshot commented:
@magento run Functional Tests B2B

monteshot commented:
@magento run Functional Tests EE

eduard13 commented on app/code/Magento/Bundle/Pricing/Price/ConfiguredPrice.php:
```diff
@@ -48,6 +48,10 @@ class ConfiguredPrice extends CatalogPrice\FinalPrice implements ConfiguredPrice
      * @var ConfiguredPriceSelection
      */
     private $configuredPriceSelection;
+    /**
```
You have missed adding an empty line above the property declaration.
```suggestion

    /**
```

monteshot replied:
I have add the missing line but github not see it

eduard13 commented on app/code/Magento/Bundle/Test/Unit/Pricing/Price/ConfiguredPriceTest.php:
```diff
@@ -0,0 +1,185 @@
+     */
+    private $productMock;
+
+    /**
+     * @var MockObject
+     */
+    private $calculator;
```
```suggestion
    /**
     * @var Calculator|MockObject
     */
    private $calculatorMock;
```

eduard13 commented on app/code/Magento/Bundle/Test/Unit/Pricing/Price/ConfiguredPriceTest.php:
```diff
@@ -0,0 +1,185 @@
+        $this->productMock->expects($this->once())->method('getPriceInfo')->willReturn($this->priceInfoMock);
+        $this->productMock->expects($this->any())->method('getId')->willReturn(123);
+
+        $this->itemMock = $this->getMockBuilder(ItemInterface::class)
+            ->getMock();
+        $this->itemMock->expects($this->any())->method('getProduct')->willReturn($this->productMock);
```
I believe it would be much better if you will move this part to `testGetValueMethod` method.

eduard13 commented on app/code/Magento/Bundle/Test/Unit/Pricing/Price/ConfiguredPriceTest.php:
```diff
@@ -0,0 +1,185 @@
+            $this->configuredPriceSelectionMock,
+            $this->discountCalculatorMock,
+        );
+        $this->model->setItem($this->itemMock);
```
Same this, should be in `testGetValueMethod`, as it represents a different case that you cover in `testGetValueMethodNoItem`

eduard13 commented on app/code/Magento/Bundle/Test/Unit/Pricing/Price/ConfiguredPriceTest.php:
```diff
@@ -0,0 +1,185 @@
+    /**
+     * @var float
+     */
+    private $basePriceValue = 100.;
```
Is there any reason of not specifying the decimals as well (and in the other parts as well)?

monteshot replied:
> Hi @monteshot, please check the below suggestions. Looks like it were missed some of the previous ones.
> Moreover, could you please provide the exact steps to reproduce the mentioned issue (maybe some piece of code)? Just to be easier to test this one.
> 
> Thank you.

This code executes on bundle product page. But i think the code in the configured price class is some type of fallback.

This code invokes in renderrer in \Magento\Catalog\Pricing\Render\FinalPriceBox block
![image](https://user-images.githubusercontent.com/2836146/125645063-3fbb489e-ec8c-4efa-a5f0-3ef5011ea0d1.png)

I did debug only if i emulate some data that will allow proceed to the if statement
![image](https://user-images.githubusercontent.com/2836146/125646776-e7837656-4e87-42bf-a67a-c6d8fd5f7720.png)
![image](https://user-images.githubusercontent.com/2836146/125646851-43fc21e8-28d1-44cf-ad96-4eb69defea26.png)

@eduard13

monteshot replied:
![image](https://user-images.githubusercontent.com/2836146/125647074-27ff67de-93bf-456a-a66a-5a7a9fc1d508.png)

monteshot replied:
> Is there any reason of not specifying the decimals as well (and in the other parts as well)?

No reason. I will add decimal after the point

monteshot replied:
Additionally you may check the original task to get steps to reproduce
https://github.com/magento/magento2/issues/33334#issue-929044829

eduard13 reviewed (changes requested):
Hi @monteshot, please check the below suggestions. Looks like it were missed some of the previous ones. 
Moreover, could you please provide the exact steps to reproduce the mentioned issue (maybe some piece of code)? Just to be easier to test this one.

Thank you.

monteshot committed:
magento/magento2#33334:Inheriting from a class that doesn't exist

- Adjusted code style for configuredPriceSelection propery in the target code
- Added mock suffice to last property in unit test
- Added decimals after point in unit test
- Transferred setting up some mocks to the specific test case

monteshot commented:
@magento run all tests

monteshot commented:
@magento run Functional Tests CE

monteshot commented:
@magento run Functional Tests CE

monteshot commented:
@magento run Functional Tests CE

monteshot committed:
magento/magento2#33334:Inheriting from a class that doesn't exist

- Added space between
$discountCalculator and $configuredPriceSelection properties
- Updated PHPdoc in constructor

monteshot commented:
@magento run all tests

monteshot commented:
@magento run Unit Tests

monteshot commented:
> Hello @monteshot
> Thank you for your contribution.
> Could you please rename your new tests?
> The tests are related with Magento\Bundle\Pricing\Price\ConfiguredPrice
> I think it should be named Magento\Bundle\Test\Unit\Pricing\Price\ConfiguredPriceTest.
> How do you think about it?

I have rename the test class to appropriate name

bgorski reviewed (approved):
Looks good to me and all the tests are passing - clicking the approve button. Thank you for your contribution!

magento-engcom-team commented:
Hi @bgorski, thank you for the review.
[ENGCOM-9161](https://jira.corp.magento.com/browse/ENGCOM-9161) has been created to process this Pull Request

magento-engcom-team commented:
Hi @Den4ik, thank you for the review.
[ENGCOM-9161](https://jira.corp.magento.com/browse/ENGCOM-9161) has been created to process this Pull Request

engcom-Hotel commented:
@magento run all tests

monteshot commented:
![image](https://user-images.githubusercontent.com/2836146/131848667-1d365c00-0aea-41b0-99e3-fdfb020aff81.png)
Another flaky test failed
@engcom-Hotel I will restart tests

monteshot commented:
@magento run Functional Tests EE
